bundles/org.openhab.core.config.discovery.mdns/src/main/java/org/openhab/core/config/discovery/mdns/internal/MDNSDiscoveryService.java

```diff
@@ -200,7 +207,17 @@ public void serviceRemoved(@NonNullByDefault({}) ServiceEvent serviceEvent) {
                 try {
                     ThingUID thingUID = participant.getThingUID(serviceEvent.getInfo());
                     if (thingUID != null) {
-                        thingRemoved(thingUID);
+                        ServiceInfo service = serviceEvent.getInfo();
+                        long gracePeriod = participant.getRemovalGracePeriodSeconds(service);
+                        if (gracePeriod <= 0) {
+                            thingRemoved(thingUID);
+                        } else {
+                            cancelRemovalTask(service);
+                            deviceRemovalTasks.put(service.getQualifiedName(), scheduler.schedule(() -> {
+                                thingRemoved(thingUID);
+                                cancelRemovalTask(service);
+                            }, gracePeriod, TimeUnit.SECONDS));
```

Comment:
Sorry for late comment, didn't catch in first round: consider extracting this to a private method (e.g. `scheduleRemovalTask`), so you would have one method for scheduling task and another method for cancelling task. Even though it would only have one usage, it should still simplify this method by making it slimmer, thus improve readability.

Reply:
Ok.